[DOC] Add an overview of how CRDs are used (#1627)
strimzi/strimzi-kafka-operator: documentation/book/con-custom-resources-example.adoc

// Module included in the following assemblies:
//
// assembly-custom-resources.adoc

[id='con-custom-resources-example-{context}']
= {Productname} custom resource example

> Each {Productname}-specific custom resource is defined within the specification of a CRD.

Reviewer: I don't really know what "defined within the specification of" means. I can guess what you mean, but I don't think it's well-known terminology.

```suggestion
Each {Productname}-specific custom resource is conforms to the schema defined by the CRD for the resource's `kind`.
```

Author: Changed to:
Each {Productname}-specific custom resource conforms to the schema defined by the CRD for the resource's `kind`.

To understand the relationship between a CRD and a custom resource, let's look at a sample of the CRD for a Kafka topic.

.Kafka topic CRD
[source,yaml,subs="attributes+"]
----
apiVersion: {KafkaApiVersion}
kind: CustomResourceDefinition
Metadata: <1>
  name: kafkatopics.kafka.strimzi.io
  labels:
    app: strimzi
Spec: <2>
  group: kafka.strimzi.io
  versions:
    {ApiVersion}
  scope: Namespaced
  names:

>     # ...

Reviewer: Please don't elide the singular name.

Author: Added

    plural: kafkatopics
    shortNames:
    - kt <3>
  additionalPrinterColumns: <4>
      # ...
  Validation: <5>
    openAPIV3Schema:
      properties:
        spec:
          type: object
          properties:
            partitions:
              type: integer
              minimum: 1
            replicas:
              type: integer
              minimum: 1
              maximum: 32767
      # ...
----

> <1> The metadata for the topic custom resource, its name and a label to identify the resource.

Reviewer: This metadata is defining a name and a label for the CRD, not a CR instance.

Author: Updated to say CRD:
The metadata for the topic CRD, its name and a label to identify the CRD.

> <2> The specification for the topic, including the group (domain) name, the plural name and the supported schema version, which are used in the URL to access the API of the topic. The other names are used to identify the resource in the CLI.

Reviewer: * This is the specification of the CRD, not a topic: "The specification for the topic" → "The specification for this CRD"
* "The other names are used to identify the resource in the CLI." → "The other names are used to identify instance resources in the CLI, for example `kubectl get kafkatopic my-topic` or `oc get kafkatopics`"

Author: Updated callout to:

<2> The specification for this CRD, including the group (domain) name, the plural name and the supported schema version, which are used in the URL to access the API of the topic. The other names are used to identify instance resources in the CLI. For example, `kubectl get kafkatopic my-topic` or `oc get kafkatopics`.

> <3> The shortname can be used in CLI commands, so `get kt` is used instead of `get kafka topic`.

Reviewer: ```suggestion
<3> The shortname can be used in CLI commands, so `get kt` can be used as an abbreviation of `get kafka topic`.
```

but really you should mention kubectl/oc because otherwise it's not quite clear what you mean.

Author: As it's an example, I've used just oc:

For example, `oc get kt` can be used as an abbreviation instead of `oc get kafka topic`.

<4> The information presented when using a `get` command on the custom resource.
<5> openAPIV3Schema validation provides validation for the creation of topic custom resources. For example, a topic requires at least one partition and one replica.

NOTE: You can identify the CRD YAML files supplied with the {Productname} installation files, as the file names contain an index number followed by ‘Crd’.

Here is a corresponding example of a `KafkaTopic` custom resource.

.Kafka topic custom resource
[source,yaml,subs="attributes+"]
----
apiVersion: {KafkaApiVersion}

> kind: KafkaTopic

Reviewer: Callout that together the `kind` and `apiVersion` identify the CRD of which this CR is an instance.

Author: New callout added.

metadata:
  name: my-topic
  labels:
    strimzi.io/cluster: my-cluster
Spec: <1>
  partitions: 1
  replicas: 1
  config:
    retention.ms: 7200000
    segment.bytes: 1073741824
----
<1> The spec shows the number of partitions and replicas for the topic as well as configuration for the retention period for a message to remain in the topic and the segment file size for the log.

> Custom resources are applied to a cluster through the platform CLI. When the custom resource is created, it uses the same validation as the built-in resources of the Kubernetes API.

Reviewer: ```suggestion
Custom resources can be applied to a cluster through the platform CLI. When the custom resource is created or modified it is validated against the CRD `openAPIV3Schema`.
```

Note: Although we use CLI tools do create CR instances that's not the _only_ way. The CLI tools are just talking REST with the API server.

Author: Adjusted

// Module included in the following assemblies:
//
// assembly-custom-resources.adoc

[id='con-custom-resources-example-{context}']

> = {Productname} custom resource example

Reviewer 2: {ProductName}

> Each {Productname}-specific custom resource conforms to the schema defined by the CRD for the resource's `kind`.

Reviewer 2: {ProductName}

.Kafka topic CRD
[source,yaml,subs="attributes+"]
----
apiVersion: {KafkaApiVersion}
kind: CustomResourceDefinition
metadata: <1>
  name: kafkatopics.kafka.strimzi.io
  labels:
    app: strimzi
spec: <2>
  group: kafka.strimzi.io
  versions:
    {ApiVersion}
  scope: Namespaced
  names:
    # ...
    singular: kafkatopic
    plural: kafkatopics
    shortNames:
    - kt <3>
  additionalPrinterColumns: <4>
      # ...
  validation: <5>
    openAPIV3Schema:
      properties:
        spec:
          type: object
          properties:
            partitions:
              type: integer
              minimum: 1
            replicas:
              type: integer
              minimum: 1
              maximum: 32767
      # ...
----
<1> The metadata for the topic CRD, its name and a label to identify the CRD.
<2> The specification for this CRD, including the group (domain) name, the plural name and the supported schema version, which are used in the URL to access the API of the topic. The other names are used to identify instance resources in the CLI. For example, `kubectl get kafkatopic my-topic` or `oc get kafkatopics`.

> <3> The shortname can be used in CLI commands. For example, `oc get kt` can be used as an abbreviation instead of `oc get kafka topic`.

Reviewer 3: explaining 2 and 3 we are mixing kubectl and oc, would it be better chosing just one for the examples?
Furthermore on point 3, it's `kafkatopic` not `kafka topic` in the oc command.

<4> The information presented when using a `get` command on the custom resource.
<5> openAPIV3Schema validation provides validation for the creation of topic custom resources. For example, a topic requires at least one partition and one replica.

> NOTE: You can identify the CRD YAML files supplied with the {Productname} installation files, as the file names contain an index number followed by ‘Crd’.

Reviewer 2: Suggested changes:

{ProductName}

as the file names contain... > because the file names contain...

Here is a corresponding example of a `KafkaTopic` custom resource.

.Kafka topic custom resource
[source,yaml,subs="attributes+"]
----
apiVersion: {KafkaApiVersion}
kind: KafkaTopic <1>
metadata:
  name: my-topic
  labels:
    strimzi.io/cluster: my-cluster
spec: <2>
  partitions: 1
  replicas: 1
  config:
    retention.ms: 7200000
    segment.bytes: 1073741824
----
<1> The `kind` and `apiVersion` identify the CRD of which the custom resource is an instance.
<2> The spec shows the number of partitions and replicas for the topic as well as configuration for the retention period for a message to remain in the topic and the segment file size for the log.

Custom resources can be applied to a cluster through the platform CLI. When the custom resource is created, it uses the same validation as the built-in resources of the Kubernetes API.

> After a KafkaTopic custom resource is created, the Topic Operator is notified and corresponding topics are created in {Productname}.

Reviewer 2: Suggested change:

corresponding topics > corresponding Kafka topics
